### PR TITLE
Rust: Sidewalk

### PR DESCRIPTION
**Pull request by droogmic**
Add sidewalk support, fixes issue #49.

### CODE REVIEW

File: rust/osm2lanes/src/transform/tags_to_lanes/foot_shoulder.rs [hunk 1]
```diff
@@ -23,75 +23,49 @@ pub fn foot_and_shoulder(
     }
     let sidewalk: (Sidewalk, Sidewalk) = match (
         tags.get(SIDEWALK),
+        tags.get(SIDEWALK + "both"),
```
[the remark this answers is not in the document]
Comment (dabreegster) [reply rust/osm2lanes/src/transform/tags_to_lanes/foot_shoulder.rs:26]: An idea for simplifying -- not sure if this is correct, or more readable. But IIUC, we have 4 mutually exclusive situations:

1) Only `sidewalk` is defined
2) Only `sidewalk:both` is defined
3) At least one of `sidewalk:left` and `sidewalk:right` are defined
4) None of them

We could say something like
```
if let Some(value) = tags.get(SIDEWALK) {
  if tags.has(SIDEWALK + "both") || left right cases {
    return Unsupported(multiple styles used)
  }
  // handle just this case
}
// etc for the other cases
```

Or actually, model what you do for the 3 busway schemas: https://github.com/a-b-street/osm2lanes/blob/d07fb6c693cd20a5167252da23add5aa4d092f29/rust/osm2lanes/src/transform/tags_to_lanes/bus.rs#L14

Detect which situation we're in up-front, handle it separately, and detect multiple situations upfront. Would any of that be more maintainable?

**Review by dabreegster — APPROVED** [review]

File: rust/osm2lanes/src/transform/tags_to_lanes/foot_shoulder.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (droogmic) [reply rust/osm2lanes/src/transform/tags_to_lanes/foot_shoulder.rs:26]: I can probably make these 4 mutually exclusive situations more clear. I'll have a go at reworking

**Comment by tordans** [comment]
Just for me to learn (and maybe as a suggestion): 

Are you normalizing tags before processing? Like renaming `sidewalk=yes` to `sidewalk:both=yes` (or the other way around)?

I image with this project, a first set of tests would be about normalization and the the second set to test on the normalized data?

**Comment by droogmic** [comment]
> Just for me to learn (and maybe as a suggestion):
> 
> Are you normalizing tags before processing? Like renaming `sidewalk=yes` to `sidewalk:both=yes` (or the other way around)?
> 
> I image with this project, a first set of tests would be about normalization and the the second set to test on the normalized data?

Depending on your definition of normalization and categorization, yes and no.
This is actually a categorizing step (a type of normalization?), the enum at https://github.com/a-b-street/osm2lanes/blob/main/rust/osm2lanes/src/transform/tags_to_lanes/foot_shoulder.rs#L18 represents all the variants of sidewalk for each side as implied by the tags. We then parse the tag strings to find which variant is valid for each side.
A later step can then additionally guess if a shoulder or sidewalk is needed based on other input.

The match statement is the core of the business logic:
```
    match (
        tags.get(SIDEWALK),
        tags.get(SIDEWALK + "both"),
        tags.get(SIDEWALK + locale.driving_side.tag()),
        tags.get(SIDEWALK + locale.driving_side.opposite().tag()),
    )
```

We are constructing a switch statement that needs to cover all values of `sidewalk`, `sidewalk:both`, `sidewalk:left` and `sidewalk:right`. The reason to do this instead of normalization where `sidewalk` is blindly combined with `sidewalk:both` is it would allow us to catch erroneous cases.

We can now handle unusual situations like this:
```
sidewalk=yes
sidewalk:both=yes
```
(we currently throw an error, but we could change it to just throw a warning)

so that ugly cases like the following are caught as an error:
```
sidewalk:both=yes
sidewalk:left=no
```

Normalization could imply mutability which is a bit of a code smell (especially in rust), so I think I will try to go as far as possible without mutating the input tags by categorizing where needed.


As an aside:
Your work in Berlin inspired me to work on this, so I hope I can bring this to a level where it is useful for your work.

**Comment by droogmic** [comment]
> test on the normalized data

We maybe do something like what you are expecting, by testing that we can roundtrip.
so if we convert tags to lanes, we check that those lanes -> tags -> lanes is a perfect roundtrip.

**Comment by droogmic** [comment]
I tried https://github.com/a-b-street/osm2lanes/commit/d0ca4b6e64c21c16d0a196c42fb569e6039cac5f

but I don't like the `(None, None) => unreachable!()`,
and I am also not sure it adds much.

will try something else

File: rust/osm2lanes/src/transform/tags_to_lanes/foot_shoulder.rs [hunk 2]
```diff
@@ -21,86 +21,69 @@ pub fn foot_and_shoulder(
...
+            _ => return err,
+        },
+        // sidewalk:both=
+        (None, Some(v), (None, None)) => match v {
+            "no" => (Sidewalk::No, Sidewalk::No),
+            "yes" => (Sidewalk::Yes, Sidewalk::Yes),
+            "separate" => (Sidewalk::Separate, Sidewalk::Separate),
+            _ => return err,
+        },
+        // sidewalk:left= and/or sidewalk:right=
```
Comment (dabreegster) [inline rust/osm2lanes/src/transform/tags_to_lanes/foot_shoulder.rs:69]: This is way more readable IMO, nice!

**Review by dabreegster — APPROVED** [review]

**Pull request merged**